File: go/4. SQLite/main.go
package main

import (
	"database/sql"
	"fmt"
	_ "github.com/mattn/go-sqlite3"
	"os"
	"strconv"
	"strings"
	"time"
)

func routine(db *sql.DB, age, id int, ch chan<- string) {
	if age%2 == 0 {
		update, _ := db.Prepare("UPDATE users SET salary = salary + 1 WHERE id = (?)")
		defer update.Close()
		update.Exec(id)

		get, _ := db.Prepare("SELECT salary FROM users WHERE id = (?)")
		defer get.Close()
		var salary string
		get.QueryRow(id).Scan(&salary)
		ch <- salary
	} else {
		rows, _ := db.Query("SELECT salary FROM users WHERE age = 10 ORDER BY id")
		defer rows.Close()

		salaries := []string{}
		for rows.Next() {
			var salary string
			rows.Scan(&salary)
			salaries = append(salaries, salary)
		}

		ch <- strings.Join(salaries, ", ")
	}
}

func lcg(n, maxValue, seed int) []int {
	timers := make([]int, n)
	timers[0] = seed

	m := 0x7FFFFFFF
	a := 48271
	c := 0
	for i := 1; i < n; i++ {
		timers[i] = (timers[i-1]*a + c) % m
		timers[i-1] = timers[i-1] % maxValue
	}
	timers[n-1] = timers[n-1] % maxValue

	return timers
}

func main() {
	n, _ := strconv.Atoi(os.Args[1])

	ages := lcg(n, 100, 42)
	ids := lcg(n, 10_000, 42)
	ch := make(chan string)
	list := make([]string, n)

	db, _ := sql.Open("sqlite3", "database.db")
	defer db.Close()

	start := time.Now()
	for i := 0; i < n; i++ {
		go routine(db, ages[i], ids[i], ch)
	}
	for i := 0; i < n; i++ {
		list[i] = <-ch
	}
	fmt.Println(time.Since(start).Microseconds())
}

File: rust/4. SQLite/src/main.rs
use std::error::Error;
use std::time::Instant;
use tokio::task::JoinHandle;
use sqlx::sqlite::SqlitePool;

async fn routine(pool: &SqlitePool, age: i32, id: i32) -> Result<String, Box<dyn Error>> {
    if age % 2 == 0 {
        sqlx::query!("UPDATE users SET salary = salary + 1 WHERE id = ?1", id)
            .execute(pool)
            .await?;
        let row = sqlx::query!("SELECT salary FROM users WHERE id = ?1", id)
            .fetch_one(pool)
            .await?;
        Ok(format!("{}: {}", id, row.salary))
    } else {
        let res = sqlx::query!("SELECT salary FROM users WHERE age = ?1 ORDER BY id", age)
            .fetch_all(pool)
            .await?
            .iter()
            .map(|it| it.salary.to_string())
            .collect::<Vec<String>>()
            .join(", ");
        Ok(res)
    }
}

fn lcg(n: usize, max_value: i128, seed: i128) -> Vec<i32> {
    let mut list: Vec<i128> = Vec::with_capacity(n);
    list.push(seed);

    let m = 0x7FFFFFFF;
    let a = 48271;
    let c = 0;
    for i in 1..n {
        list.push((list[i - 1] * a + c) % m);
        list[i - 1] = list[i - 1] % max_value;
    }
    list[n - 1] = list[n - 1] % max_value;

    list.iter().map(|it| *it as i32).collect()
}

#[tokio::main]
async fn main() {
    let n = std::env::args().nth(1).unwrap().parse().unwrap();
    let pool = SqlitePool::connect("sqlite://database.db").await.unwrap();
    let ages = lcg(n, 100, 42);
    let ids = lcg(n, 10_000, 17);

    let start = Instant::now();
    let handles: Vec<JoinHandle<_>> = (0..n).map(|i| {
        let pool = pool.clone();
        let age = ages[i];
        let id = ids[i];
        tokio::spawn(async move {
            routine(&pool, age, id).await.unwrap()
        })
    }).collect();

    for handle in handles {
        handle.await.unwrap();
    }

    println!("{:?}", start.elapsed().as_micros());
}
